src/network/core/tcp.h

```diff
@@ -132,6 +134,18 @@ class TCPConnecter {
 
 	static void CheckCallbacks();
 	static void KillAll();
+
+	/**
+	 * Create the connecter, and initiate connecting by putting it in the collection of TCP connections to make.
+	 * @tparam T The type of connecter to create.
+	 * @param args The arguments to the constructor of T.
+	 * @return Non-owning pointer to T.
+	 */
+	template <class T, typename... Args>
+	static T *Create(Args&& ... args)
+	{
+		return static_cast<T *>(TCPConnecter::connecters.emplace_back(std::make_unique<T>(std::forward<Args>(args)...)).get());
```

Comment:
This feels so wrong to do ... this is heavily abusing what an unique-ptr stands for. And I am rather unsure if this is actually an improvement.

I would say, we could better make it a shared-ptr, and also change things like `this->game_connecter` to be a shared-ptr.

But to have an unique-ptr, and steal its pointer to give to other to do what-ever they want with it .. I am not sure we actually solving any problems here :) Mostly, and this has been a problem, it doesn't solve that `this->game_connecter` still has a pointer while the object is already deleted.

Reply:
There once was the idea of [`std::observer_ptr`](https://www.open-std.org/jtc1/sc22/wg21/docs/papers/2019/p1408r0.pdf), which is what this function would report. But that has been killed off, and `T*` is the de-facto standard for non-owning pointers. That's exactly what it's returning here.
So, I'm not making a `unique_ptr` to steal it's pointer, it's giving a pointer to observe the object.

What used to happen was creating the connecter and secretly (in the constructor) the ownership of the object was hijacked. The network coordinator was not allowed to delete the `game_connecter`, even though it created the object itself. So in that sense, the new code is better as the owner gives an observer pointer to the network coordinator.

In all code paths where the `unique_ptr` gets deleted, there is first a call of some sort that sets `game_connecter` to `nullptr`. So, there is no problem.

Having said that, I do agree that `std::shared_ptr` is the safer option in the long run.